egs/librispeech/ASR/pruned_transducer_stateless2/scaling.py

```diff
@@ -111,6 +111,74 @@ def backward(
         return x_grad - neg_delta_grad, None, None, None, None, None, None
 
 
+class GradientFilterFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        x: Tensor,
+        batch_dim: int,  # e.g., 1
+        threshold: float,  # e.g., 10.0
+        *params: nn.Parameter,  # module parameters
+    ):
+        if x.requires_grad:
+            if batch_dim < 0:
+                batch_dim += x.ndim
+            ctx.batch_dim = batch_dim
+            ctx.threshold = threshold
+        return x, *params
+
+    @staticmethod
+    def backward(
+        ctx,
+        x_grad: Tensor,
+        *param_grads: nn.Parameter,
+    ):
+        dim = ctx.batch_dim
+        norm_dims = [d for d in range(x_grad.ndim) if d != dim]
+        norm_of_batch = (x_grad ** 2).mean(dim=norm_dims, keepdim=True).sqrt()
+        median_norm = norm_of_batch.median()
+        mask = norm_of_batch <= median_norm * ctx.threshold
+        x_grad = x_grad * mask
+
+        if not torch.all(mask):
+            # If any of elements of x_grad is zeroed,
+            # the param_grads would be fully zeroed.
+            for g in param_grads:
+                g.zero_()
+        return x_grad, None, None, *param_grads
+
+
+class GradientFilter(torch.nn.Module):
+    """This is used to filter out elements that have extremely large gradients
+    in batch.
+
+    Args:
+      batch_dim (int):
+        The batch dimension.
+      threshold (float):
+        For each element in batch, its gradient will be
+        filtered out if the gradient norm is larger than
+        `grad_norm_threshold * median`, where `median` is the median
+        value of gradient norms of all elememts in batch.
+    """
+
+    def __init__(self, batch_dim: int = 1, threshold: float = 10.0):
+        super(GradientFilter, self).__init__()
+        self.batch_dim = batch_dim
+        self.threshold = threshold
+
+    def forward(self, x: Tensor, *params: nn.Parameter) -> Tensor:
+        if torch.jit.is_scripting() or is_jit_tracing():
```

Comment:
fix return type.  can just be Tuple[Tensor] probably, but not sure if typing expets that to be just one tensor.